**Numerical Linear Algebra for Financial Engineering**

**Homework 1**

**Jose Ferreira**

----------------
**(1) Find the value of $x$ such that the matrix below has rank 3:**
$$
   \begin{pmatrix} 
      x & 1 & 1 & 1 \\ 
      1 & x & 1 & 1 \\ 
      1 & 1 & x & 1 \\
      1 & 1 & 1 & x 
   \end{pmatrix}
$$

Let M be the matrix above. Calculate the determinant of M by applying row reducing operations until the matrix is upper triangular. 
For the rank to be 3 (or less), the determinant needs to be equal to zero.
$$
\begin{align}
det(M)=&
   -\begin{vmatrix} 
      1 & x & 1 & 1 \\ 
      x & 1 & 1 & 1 \\ 
      1 & 1 & x & 1 \\
      1 & 1 & 1 & x 
   \end{vmatrix} &[swap(r1,r2)]\\ 
=&
   -\begin{vmatrix} 
      1 & x & 1 & 1 \\ 
      0 & 1-x^2 & 1-x & 1-x \\ 
      1 & 1 & x & 1 \\
      1 & 1 & 1 & x 
   \end{vmatrix} &[r2 = r2 - x*r1]\\
=&
   -\begin{vmatrix} 
      1 & x & 1 & 1 \\ 
      0 & 1-x^2 & 1-x & 1-x \\ 
      0 & 1-x & x-1 & 0 \\
      1 & 1 & 1 & x 
   \end{vmatrix} &[r3 = r3 - r1]\\
=&
   -\begin{vmatrix} 
      1 & x & 1 & 1 \\ 
      0 & 1-x^2 & 1-x & 1-x \\ 
      0 & 1-x & x-1 & 0 \\
      0 & 1-x & 0 & x-1 
   \end{vmatrix} &[r4 = r4 - r1]\\
=&
   -\begin{vmatrix} 
      1 & x & 1 & 1 \\ 
      0 & 1-x^2 & 1-x & 1-x \\ 
      0 & 0 & \frac{(x-1)(2+x)}{1+x} & \frac{x-1}{1+x} \\
      0 & 1-x & 0 & x-1 
   \end{vmatrix} &[r3 = r3 -\frac{1-x}{1-x^2}*r2]\\   
=&
   -\begin{vmatrix} 
      1 & x & 1 & 1 \\ 
      0 & 1-x^2 & 1-x & 1-x \\ 
      0 & 0 & \frac{(x-1)(2+x)}{1+x} & \frac{x-1}{1+x} \\
      0 & 0 & \frac{x-1}{1+x} & \frac{(x-1)(2+x)}{1+x} 
   \end{vmatrix} &[r4 = r4 -\frac{1-x}{1-x^2}*r2]\\   
=&
   -\begin{vmatrix} 
      1 & x & 1 & 1 \\ 
      0 & 1-x^2 & 1-x & 1-x \\ 
      0 & 0 & \frac{(x-1)(2+x)}{1+x} & \frac{x-1}{1+x} \\
      0 & 0 & 0 & \frac{(x-1)(3+x)}{2+x} 
   \end{vmatrix} &[r4 = r4 -\frac{1}{2+x}*r3]\\      
=& -\frac{(1-x^2)(x-1)(2+x)(x-1)(3+x)}{(1+x)(2+x)}\\
=& (x-1)^3(3+x)
\end{align}
$$

Therefore, in order for the matrix to have a rank of 3 or less, $x$ must be $1$  or $-3$. If $x = 1$, the matrix is composed of only ones and its rank would be $1$. If $x = -3$:

$$
\begin{align}
M=&   \begin{pmatrix} 
      -3 & 1 & 1 & 1 \\ 
      1 & -3 & 1 & 1 \\ 
      1 & 1 & -3 & 1 \\
      1 & 1 & 1 & -3 
   \end{pmatrix}\\
=&
   \begin{pmatrix} 
      1 & -3 & 1 & 1 \\ 
      -3 & 1 & 1 & 1 \\ 
      1 & 1 & -3 & 1 \\
      1 & 1 & 1 & -3 
   \end{pmatrix} &[swap(r1,r2)]\\ 
=&
   \begin{pmatrix} 
      1 & -3 & 1 & 1 \\ 
      0 & -8 & 4 & 4 \\ 
      1 & 1 & -3 & 1 \\
      1 & 1 & 1 & -3 
   \end{pmatrix} &[r2=r2+3r1]\\ 
=&
   \begin{pmatrix} 
      1 & -3 & 1 & 1 \\ 
      0 & -8 & 4 & 4 \\ 
      0 & 4 & -4 & 0 \\
      1 & 1 & 1 & -3 
   \end{pmatrix} &[r3=r3-r1]\\ 
=&
   \begin{pmatrix} 
      1 & -3 & 1 & 1 \\ 
      0 & -8 & 4 & 4 \\ 
      0 & 4 & -4 & 0 \\
      0 & 4 & 0 & -4 
   \end{pmatrix} &[r4=r4-r1]\\ 
=&
   \begin{pmatrix} 
      1 & -3 & 1 & 1 \\ 
      0 & -8 & 4 & 4 \\ 
      0 & 0 & -2 & 2 \\
      0 & 4 & 0 & -4 
   \end{pmatrix} &[r3=r3+\frac{1}{2}r2]\\
=&
   \begin{pmatrix} 
      1 & -3 & 1 & 1 \\ 
      0 & -8 & 4 & 4 \\ 
      0 & 0 & -2 & 2 \\
      0 & 0 & 2 & -2 
   \end{pmatrix} &[r4=r4+\frac{1}{2}r2]\\   
=&
   \begin{pmatrix} 
      1 & -3 & 1 & 1 \\ 
      0 & -8 & 4 & 4 \\ 
      0 & 0 & -2 & 2 \\
      0 & 0 & 0 & 0 
   \end{pmatrix} &[r4=r4+r3]\\   
\iff& rank(M) = 3 
\end{align}
$$

-------------
**(2) Show that the product of two symmetric matrices is not necessarily a symmetric matrix.
Hint: Find 2x2 matrices $A$ and $B$ such that $A=A^t$, $B=B^t$, and $(AB)^t  \neq AB$**

By counterexample, let A and B be matrices of the form:
    $$
    A=\begin{pmatrix} 
      a & a \\ 
      a & -a
   \end{pmatrix}\\
   B=\begin{pmatrix} 
      0 & a \\ 
      a & -a
   \end{pmatrix}\\
    $$

Both A and B are symmetric matrices but:
$$
AB=\begin{pmatrix} 
      a^2 & 0 \\ 
      -a^2 & 2a^2
   \end{pmatrix}\\
$$
is not symmetric

---------------
**(3) Let $A$ be a square matrix of size n. Show that $AD=DA$ for any diagonal matrix D of size n if and only if $A$ is a diagonal matrix.**

Let $D = diag(d_{k})_{k=1:n}$ and $A = col(a_{k})_{k=1:n}$, then $AD = col(a_{k}d_{k}) \iff (AD)(j,k) = d_{k}A(j,k)$

Also, if $A = row(r_{j})_{j=1:n}$, then $DA = row(d_{j}r_{j})_{j=1:n} \iff (DA)(j,k) = d_{j}A(j,k)$

Assuming all the diagonal entries in D are distinct: 

$$
\begin{align}
AD = DA 
\iff& d_{k}A(j,k) = d_{j}A(j,k)\\ 
\iff& (d_{k} - d_{j})A(j,k) = 0, \forall 1\leq j \neq k \leq n\\
\iff& A(j,k) = 0, \forall 1\leq j \neq k \leq n\\
\iff& A = diag(a_{k})_{k=1:n}
\end{align}
$$

If the diagonal entries of D are not all distinct, some non-diagonal entries of A could be non-zero and the commutative property will also be preserved.

-------------
**(4) Let $M_1$, $M_2$, $M_3$ and $M_4$ be square matrices of the same size such that $M_1M_2M_3M_4 = I$.
Show that
$$M_1M_2M_3M_4 = M_2M_3M_4M_1 = M_3M_4M_1M_2 = M_4M_1M_2M_3 = I$$**

$$
\begin{align}
M_1M_2M_3M_4 = I \implies& M_1(M_2M_3M_4) = I\\
\implies& (M_2M_3M_4) = M_1^{-1}\\
\implies& (M_2M_3M_4)M_1 = M_2M_3M_4M_1 = M_1^{-1}M_1 = I\\
\end{align}
$$

$$
\begin{align}
M_1M_2M_3M_4 = I \implies& (M_1M_2)(M_3M_4) = I\\
\implies& (M_3M_4) = (M_1M_2)^{-1}\\
\implies& (M_3M_4)(M_1M_2) = M_3M_4M_1M_2 = (M_1M_2)^{-1}(M_1M_2) = I\\
\end{align}
$$

$$
\begin{align}
M_1M_2M_3M_4 = I \implies& (M_1M_2M_3)M_4 = I\\
\implies& M_4 = (M_1M_2M_3)^{-1}\\
\implies& (M_4)(M_1M_2M_3) = M_4M_1M_2M_3 = (M_1M_2M_3)^{-1}(M_1M_2M_3) = I\\
\end{align}
$$

---------
**(5) If A is a nonsingular matrix, show that
$$(I+A)^{-1} = I - (I+A^{-1})^{-1}$$**

$$
\begin{align}
(I+A)^{-1} = I - (I+A^{-1})^{-1}\\ 
\implies& (I - (I+A^{-1})^{-1})(I+A) = I &\text{[def. inverse matrix]}\\
\implies& I - (I+A^{-1})^{-1} + A - (I+A^{-1})^{-1}A = I &\text{[distributive property]}\\
\implies& I - A[(I+A^{-1})^{-1}A^{-1} + I - (I+A^{-1})^{-1}] = I &\text{[factor A]}\\
\implies& I - A[(I+A^{-1})^{-1}(I+A^{-1}) - I] = I &\text{[factor } (I+A)^{-1} \text{]} \\
\implies& I - A[I - I] = I &\text{[simplify]} \\
\implies& I = I &\text{[simplify]} \\
\implies& TRUE \\
\end{align}
$$

-------------
**(6) Show that the $n \times n$ matrix $A$ has rank 1 if and only if there exists an $n \times 1$ nonzero vector $v$ such that $A=vv^t$** 

If $A=vv^t$, all rows are a linear combination of vector $v$ and, therefore, $rank(A) = 1$. Also, note that a vector has by definition a rank equal to 1, multiplying two vectors must result in a matrix of rank equal to 1. 

$$
\begin{align}
A=vv^t =& (vv^t)(j,k) = v_{j}v^t_{k}, \forall j =1:n, k=1:n\\
\implies& Ax = (v_{j}row(r_{j}))x, \forall j =1:n\\
\implies& Ax = v_{j}(row(r_{j})x), \forall j =1:n\\
\implies& \text{All rows are linearly dependent}\\
\implies& rank(A) = 1
\end{align}
$$

If $rank(A) = 1$, there exists exactly one eigenvector $v$ such that $Av = \lambda v$ and, assuming $v$ is a nonzero vector, the matrix can be expressed as $A = vv^t$ since any vector $w$ such that $A = vw^t$ (possible since rank(A) = 1) can be written as $w = \lambda v$ if  $\vert\vert{v}\vert\vert=\sqrt{\lambda}$.      

$$
\begin{align}
rank(A) = 1\\
\implies& Au = kv, \forall u \in \mathbb R^n \\
\implies& A = v(col(w_{k})), \forall k =1:n\\
\implies& A = vw^t\\    
\implies& A = v_*(v_*^t), v_*=\sqrt{\lambda}v \text{ since } Av = \lambda v\\
\end{align}
$$


-------------
**(7) Let $x$ and $y$ be column vectors of size $n$, and let $I$ be the identity matrix of size $n$. Show that the matrix $I - xy^t$ is nonsingular if and only if $y^tx \neq 1$.**

Let $A = (I-xy^t)$ and assume $y^tx \neq 1$:

$$
\begin{align}
(I-xy^t)A^{-1} = I\\
\implies& A^{-1} - xy^tA^{-1} = I\\
\implies& y^tA^{-1} - y^txy^tA^{-1} = y^t &\text{[multiply by $y^t$]}\\
\implies& (1 - y^tx)(y^tA^-1) = y^t\\
\implies& (1 - y^tx)(y^tA^-1) = y^t\\
\implies& A^-1 = I + \frac{xy^t}{(1 - y^tx)}\\
\end{align}
$$
Since $y^tx \neq 1$, the inverse of matrix A is well defined.  
If $y^tx = 1$, the inverse does not exist.

Furthermore, assuming $y^tx = 1$:
$$
\begin{align}
(I-xy^t)x =& x - xy^tx\\
=& x - x(y^tx)\\
=&0\\
\implies& \text{The matrix is singular if $y^tx = 1$}
\end{align}
$$

Therefore, $I - xy^t$ is nonsingular $\iff y^tx \neq 1$  

---------------
**(8) The covariance matrix of five random variables is
$$
\Sigma =
\begin{pmatrix} 
      1 & -0.525 & 1.375 & -0.075 & -0.75 \\ 
      -0.525 & 4 & 0.1875 & 0.1875 & -0.675 \\ 
      1.375 & 0.1875 & 12.25 & 0.4375 & -1.875 \\ 
      -0.075 & 0.1875 & 0.4375 & 6.25 & 0.3 \\ 
      -0.75 & -0.675 & -1.875 & 0.3 & 4.41
   \end{pmatrix}\\
$$
Find the correlation matrix of these random variables.** 

The correlation matrix $\Omega$ is given by
$$\Omega = (D_{\sigma})^{-1}\Sigma(D_{\sigma})^{-1}$$ 
where $D_{\sigma}$ is the diagonal matrix with diagonal values equivalent to the standard deviations of each random variable $X_{i}$  
In this case, 
$$
\begin{align}
D_{\sigma} =& 
\begin{pmatrix} 
      \sqrt{1} & 0 & 0 & 0 & 0 \\ 
      0 & \sqrt{4} & 0 & 0 & 0 \\ 
      0 & 0 & \sqrt{12.25} & 0 & 0 \\ 
      0 & 0 & 0 & \sqrt{6.25} & 0 \\ 
      0 & 0 & 0 & 0 & \sqrt{4.41}
   \end{pmatrix}\\
=& 
\begin{pmatrix} 
      1 & 0 & 0 & 0 & 0 \\ 
      0 & 2 & 0 & 0 & 0 \\ 
      0 & 0 & 3.5 & 0 & 0 \\ 
      0 & 0 & 0 & 2.5 & 0 \\ 
      0 & 0 & 0 & 0 & 2.1
\end{pmatrix}\\
\end{align}
$$

$$
\begin{align}
(D_{\sigma})^{-1} 
=& diag\left(\frac{1}{d_{k}}\right)_{k=1:n}\\
=& 
\begin{pmatrix} 
      1 & 0 & 0 & 0 & 0 \\ 
      0 & \frac{1}{2} & 0 & 0 & 0 \\ 
      0 & 0 & \frac{2}{7} & 0 & 0 \\ 
      0 & 0 & 0 & \frac{2}{5} & 0 \\ 
      0 & 0 & 0 & 0 & \frac{10}{21}
\end{pmatrix}\\
\end{align}
$$

In [4]:
sigma = [1 -0.525 1.375 -0.075 -0.75; 
         -0.525 4 0.1875 0.1875 -0.675;
         1.375 0.1875 12.25 0.4375 -1.875;
         -0.075 0.1875 0.4375 6.25 0.3;
         -0.75 -0.675 -1.875 0.3 4.41]
d_sigma_inverse = inv(Diagonal([sqrt(sigma[1,1]); 
                                sqrt(sigma[2,2]);
                                sqrt(sigma[3,3]);
                                sqrt(sigma[4,4]);
                                sqrt(sigma[5,5])]))
omega = d_sigma_inverse*sigma*d_sigma_inverse

5×5 Array{Float64,2}:
  1.0       -0.2625      0.392857   -0.03       -0.357143 
 -0.2625     1.0         0.0267857   0.0375     -0.160714 
  0.392857   0.0267857   1.0         0.05       -0.255102 
 -0.03       0.0375      0.05        1.0         0.0571429
 -0.357143  -0.160714   -0.255102    0.0571429   1.0      

$$
\begin{align}
\Omega
=& 
\begin{pmatrix} 
1.0 & -0.2625 & 0.392857 & -0.03 & -0.357143 \\ 
-0.2625 & 1.0 & 0.0267857 & 0.0375 & -0.160714 \\
0.392857 & 0.0267857 & 1.0 & 0.05 & -0.255102 \\
-0.03 & 0.0375 & 0.05 & 1.0 & 0.0571429 \\
-0.357143 & -0.160714 & -0.255102 & 0.0571429 & 1.0   
\end{pmatrix}\\
\end{align}
$$


----------
**(9) The correlation matrix of five random variables is
$$
\begin{align}
\Omega
=& 
\begin{pmatrix} 
1 & -0.25 & 0.15 & -0.05 & -0.3 \\ 
-0.25 & 1 & -0.1 & -0.25 & 0.1\\
0.15 & -0.1 & 1 & 0.2 & 0.05 \\
-0.05 & -0.25 & 0.2 & 1 & 0.1 \\
-0.3 & 0.1 & 0.05 & 0.1 & 1   
\end{pmatrix}\\
\end{align}
$$
(i) Compute the covariance matrix of these random variables if their standard deviations are 0.1, 0.2, 0.5, 1 and 2, in this order.**

The covariance matrix, $\Sigma$ is given by the expression $$\Sigma = D_{\sigma}\Omega D_{\sigma}$$ where 
$$
\begin{align}
D_{\sigma} =& diag\left(\sigma_{k}\right)_{k=1:n}\\
=&
\begin{pmatrix} 
      0.1 & 0 & 0 & 0 & 0 \\ 
      0 & 0.2 & 0 & 0 & 0 \\ 
      0 & 0 & 0.5 & 0 & 0 \\ 
      0 & 0 & 0 & 1 & 0 \\ 
      0 & 0 & 0 & 0 & 2
\end{pmatrix}\\
\end{align}
$$ 


In [8]:
omega = [1 -0.25 0.15 -0.05 -0.3; 
         -0.25 1 -0.1 -0.25 0.1;
         0.15 -0.1 1 0.2 0.05;
         -0.05 -0.25 0.2 1 0.1;
         -0.3 0.1 0.05 0.1 1]
d_sigma = Diagonal([0.1; 0.2; 0.5; 1; 2])
sigma = d_sigma*omega*d_sigma

5×5 Array{Float64,2}:
  0.01    -0.005   0.0075  -0.005  -0.06
 -0.005    0.04   -0.01    -0.05    0.04
  0.0075  -0.01    0.25     0.1     0.05
 -0.005   -0.05    0.1      1.0     0.2 
 -0.06     0.04    0.05     0.2     4.0 

$$
\Sigma =
\begin{pmatrix} 
      0.01 & -0.005 & 0.0075 & -0.005 & -0.06 \\ 
      -0.005 & 0.04 & -0.01 & -0.05 & 0.04 \\ 
      0.0075 & -0.01 & 0.25 & 0.1 & 0.05 \\ 
      -0.005 & -0.05 & 0.1 & 1 & 0.2 \\ 
      -0.06 & 0.04 & 0.05 & 0.2 & 4
   \end{pmatrix}\\
$$

**(ii) Compute the covariance matrix of these random variables if their standard deviations are 2, 1, 0.5, 0.2, and 0.1, in this order.**

In [9]:
omega = [1 -0.25 0.15 -0.05 -0.3; 
         -0.25 1 -0.1 -0.25 0.1;
         0.15 -0.1 1 0.2 0.05;
         -0.05 -0.25 0.2 1 0.1;
         -0.3 0.1 0.05 0.1 1]
d_sigma = Diagonal([2; 1; 0.5; 0.2; 0.1])
sigma = d_sigma*omega*d_sigma

5×5 Array{Float64,2}:
  4.0   -0.5    0.15    -0.02   -0.06  
 -0.5    1.0   -0.05    -0.05    0.01  
  0.15  -0.05   0.25     0.02    0.0025
 -0.02  -0.05   0.02     0.04    0.002 
 -0.06   0.01   0.0025   0.002   0.01  

$$
\Sigma =
\begin{pmatrix} 
      4 & -0.5 & 0.15 & -0.02 & -0.06 \\ 
      -0.5 & 1 & -0.05 & -0.05 & 0.01 \\ 
      0.15 & -0.05 & 0.25 & 0.2 & 0.0025 \\ 
      -0.02 & -0.05 & 0.02 & 0.04 & 0.002 \\ 
      -0.06 & 0.01 & 0.0025 & 0.002 & 0.01
   \end{pmatrix}\\
$$

------------
**(10) The file *indeces-close-jan3-jan31-2017.xslx* contains the January 3, 2017 - January 31, 2017 end of day values of Dow Jones, Nasdaq, and S&P 500.**

**(i) Compute the log daily returns of the three indices over the given time period.** 

**(ii) Compute the sample covariance matrix of the log daily returns of the three indices over the given time period.**

**(iii) Compute the percentage daily returns of the three indices over the given time period.**

**(iv) Compute the sample covariance matrix of the percentage daily returns of the three indices over the given time period.**

--------------
**(11) The file *indices-july2016.csv* contains the January 2016 - July 2016 end of day values of nine major US indices.**

**(i) Compute the sample covariance matrix of the daily percentage returns of the indices, and the corresponding sample correlation matrix. Compute the sample covariance and correlation matrices for daily log returns, and compare them with the corresponding matrices for daily percentage returns.**

**(ii) Compute the sample covariance matrix of the weekly percentage returns of the indices, and the corresponding sample correlation matrix. Compute the sample covariance and correlation matrices for weekly log returns, and compare them with the corresponding matrices for weekly percentage returns.**

**(iii) Compute the sample covariance matrix of the monthly percentage returns of the indices, and the corresponding sample correlation matrix. Compute the sample covariance and correlation matrices for monthly log returns, and compare them with the corresponding matrices for monthly percentage returns.**

**(iv) Comment on the differences between the sample covariance and correlation matrices for daily, weekly and monthly returns.** 

--------------------
**(12) Consider a market made of cash and an asset with spot price 40. The continuously compounded risk-free rate is 2%. In four months, the price of the stock will be either 45 or 35.**

**(i) What is the payoff matrix M in three months?**

The future value of cash is $e^{0.02\times 0.25} = 1.00501252$ and

$$
M=\begin{pmatrix} 
      1.00501252 & 1.00501252 \\ 
      45 & 35
   \end{pmatrix}\\
$$

**(ii) Is the market complete, i.e., is the matrix M nonsingular?**

$det(M) =  1.00501252\times 35 - 1.00501252\times 45 = 35.1754382 - 45.2255634 = -10.0501252 \neq 0$

$\implies$ the matrix M is nonsingular and the market is complete.

**(iii) How do you replicate a four months at-the-money call option on this asset, using the cash and the underlying asset?**